### ISSUE TITLE
Invalid TX generated by v2.9.3 (Non-canonical DER signature)

### ISSUE THREAD

**Issue by SomberNight** [spesmilo/electrum#2857]
User "Xeron" on #electrum IRC (on freenode) retrieved a bitcoin private key from an old (2010) wallet.dat using Bitcoin Core 0.14.2.
He then tried to sweep the coins from the privkey using Electrum 2.9.3, using several different servers.
He also tried creating a standard wallet directly from the privkey.

In all cases, when trying to broadcast the signed transaction, he was presented with the error
```
error: The transaction was rejected by network rules. (16: mandatory-script-verify-flag-failed (Non-canonical DER signature))
```

Unsigned TX:
```
0100000002ccae9778042adb58ac388aa0a2463eb0e68ba9e34dfcca115706510e54458b38010000004401ff41049fd7cfe066a814030c538d1ef4cbc412dec86c0215c364bca46ed267165f3cfde66e55db5738bd781ab84f69085a2c27909f20afaeae65f1d341944c5ded3c60fdffffff11e5c3079d60db1a5496bc98ad4d7fe884bae7a20eac8cb2941664bd70546697000000004401ff41046619ac0ae726b306297d85bd5e877d7e9a9922094db131297dc7e3d16b93e6e01620b57c503439a0f03d39735e9fc56c7d2e644ba963f43451b4e47306947e4afdffffff018cdfd2470100000017a91479ef7e9aec1e5943862790344d23843f41b109de8700000000
```

Signed TX:
```
0100000002ccae9778042adb58ac388aa0a2463eb0e68ba9e34dfcca115706510e54458b38010000008b483045022100dd80de2eaad8dea0f4438409483c1321930c932812dd695d39080554c0ce8243022037186ac0d340b53f81188c5cfdf53201a6fbeec192ef3e3c5190e3cb93f582f20141049fd7cfe066a814030c538d1ef4cbc412dec86c0215c364bca46ed267165f3cfde66e55db5738bd781ab84f69085a2c27909f20afaeae65f1d341944c5ded3c60fdffffff11e5c3079d60db1a5496bc98ad4d7fe884bae7a20eac8cb2941664bd70546697000000008a47304402206c51c939ee2b4b46422c6e7e65fd3f8f474f15a710366a5e96d75e17f91ba7be02204b0ade7182873b8b4e5571a797f8393e82e9fac05a76e4c7a963d6b34cc6cf3f0141046619ac0ae726b306297d85bd5e877d7e9a9922094db131297dc7e3d16b93e6e01620b57c503439a0f03d39735e9fc56c7d2e644ba963f43451b4e47306947e4afdffffff018cdfd2470100000017a91479ef7e9aec1e5943862790344d23843f41b109de8700000000
```

**Comment by dabura667** [comment]
```
3045022100 dd80de2eaad8dea0f4438409483c1321930c932812dd695d39080554c0ce8243 0220 37186ac0d340b53f81188c5cfdf53201a6fbeec192ef3e3c5190e3cb93f582f2 01
30440220   6c51c939ee2b4b46422c6e7e65fd3f8f474f15a710366a5e96d75e17f91ba7be 0220 4b0ade7182873b8b4e5571a797f8393e82e9fac05a76e4c7a963d6b34cc6cf3f 01
```
The two signatures look well formatted to me... so "Non-canonical DER signature" is not the problem.

The only other explanation is that the signature messed up somehow. Signing the wrong hash or something

**Comment by fivepiece** [comment]
Copying my message from irc:
```
<arubi> looks like the second input in Xeron's tx is spending a p2sh, but really it's a p2pk from coinbase output
<arubi> err, spending a p2pkh
<arubi> (when it's a p2pk from coinbase)
<arubi> a person was on #bitcoin a while ago that had the same difficulty spending an old p2pk with electrum.  eventually they used core
<arubi> the signature for the first input is correct but not for the second input, so seems that electrum actually tries to sign a p2pkh and not the utxo's scriptpubkey (so not just a wrong push of scriptsig)
```

**Comment by dabura667** [comment]
tl;dr

Electrum improperly signed a sweeped (or imported) P2PK output as P2PKH.

Interesting...

**Comment by qirtaiba** [comment]
My bug #3002 was merged with this one but my problem doesn't involve a P2PK output as far as I know. It's an ordinary P2PKH address that I am trying to sweep.

**Comment by dabura667** [comment]
Post the raw transaction

**Comment by qirtaiba** [comment]
Here's an example (I'm assuming that people could sweep my coins with this information, so it's got only a few satoshis in it):

{
    "complete": true,
    "final": true,
    "hex": "0100000001550206d45a254ce039d51b23db2e1e9a890b9548575bf8e24600ec51ce2a1864000000006b4830450221009ad83ae43d5b93fde14ebefc1278de3ff9359d8b82391bca23bae19d22f85e1f0220181e9774632a657a22e2e69d839ee8e70d6fa6342a9d5047154d6e43b2383c220121023c68bea0b6135bc42d6d65f89efecc49f36cc8bb5a99a2e7df9d7b108d9beac5feffffff0195230000000000001976a914952bf590dc7e0eb1f1b7b89e67d9ef5c12aeeab188ac00000000"
}

**Comment by SomberNight** [comment]
Thanks for that @qirtaiba! Let's continue discussion in #3002.
Note that publishing a signed transaction does NOT compromise your security normally, though it has privacy implications.
